### PR TITLE
[fa] mint/burn events

### PR DESCRIPTION
**Pull request by lightmark**
- Add burn/mint events for fa
- Add `BurnStore` event for secondary store deletion

All events gated by feature flag.

### CODE REVIEW

File: aptos-move/framework/aptos-framework/sources/fungible_asset.move [hunk 1]
```diff
@@ -215,6 +215,20 @@ module aptos_framework::fungible_asset {
...
+    #[event]
+    /// Emitted when fungible assets are minted.
+    struct Mint has drop, store {
+        metadata: address,
+        amount: u64,
+    }
+
+    #[event]
+    /// Emitted when fungible assets are burned.
+    struct Burn has drop, store {
```
Comment (bowenyang007) [inline aptos-move/framework/aptos-framework/sources/fungible_asset.move:227]: we don't log if the entire Store object is burned right? should we include where the burn comes from? the object address of Store?

Comment (rtso) [reply aptos-move/framework/aptos-framework/sources/fungible_asset.move:227]: also need object address and owner and whether it's primary or secondary store

File: aptos-move/framework/aptos-framework/sources/fungible_asset.move [hunk 2]
```diff
@@ -887,6 +905,10 @@ module aptos_framework::fungible_asset {
             metadata,
             amount
         } = fa;
+        event::emit(Burn {
```
Comment (bowenyang007) [inline aptos-move/framework/aptos-framework/sources/fungible_asset.move:908]: why are we emitting this in burn_internal? Is this legit?
>     /// CAN ONLY BE CALLED BY coin.move for migration.

**Review by bowenyang007 — CHANGES_REQUESTED** [review]
Based on the comments we're only calling these for functions to be used in migration. I think that it's fine for APT but what about a custom FA? Are these comments still legit?

> /// CAN ONLY BE CALLED BY coin.move for migration.

**Comment by rtso** [comment]
Can we actually create a new event for when a secondary store gets burnt that includes the store_address, previous_owner, asset_type?

File: aptos-move/framework/aptos-framework/sources/fungible_asset.move (hunk 1 above)
Comment (rtso) [reply aptos-move/framework/aptos-framework/sources/fungible_asset.move:227]: Yes we need to include object address of the store.

**Comment by JohnChangUK** [comment]
An event for deleting/removing stores would be beneficial too. **[here](https://github.com/aptos-labs/aptos-core/blob/main/aptos-move/framework/aptos-framework/sources/fungible_asset.move#L756)**

File: aptos-move/framework/aptos-framework/sources/fungible_asset.move (hunk 2 above)
Comment (JohnChangUK) [reply aptos-move/framework/aptos-framework/sources/fungible_asset.move:908]: This simply means the `coin` module is only allowed to call this function, as it is `(friend)` gated.

File: aptos-move/framework/aptos-framework/sources/fungible_asset.move (hunk 1 above)
Comment (JohnChangUK) [reply aptos-move/framework/aptos-framework/sources/fungible_asset.move:227]: Synced with @rtso , this `Burn` event is sufficient as there is no way to know which store the FAs are being burned from this function: `public fun burn(ref: &BurnRef, fa: FungibleAsset)`.

**Comment by lightmark** [comment]
> Based on the comments we're only calling these for functions to be used in migration. I think that it's fine for APT but what about a custom FA? Are these comments still legit?
> 
> > /// CAN ONLY BE CALLED BY coin.move for migration.

It means besides this module itself, it will be only called by coin.move for migration purpose. And it applies to all the coin types.

**Review by bowenyang007 — APPROVED** [review]

**Review by davidiw — CHANGES_REQUESTED** [review]
if we leave this on for apt, we'll end up with a lot of burn and mint events... this will have a performance impact for low roi. have we walked through this? do we want to disable apt at a later date?

File: aptos-move/framework/aptos-framework/sources/fungible_asset.move [hunk 3]
```diff
@@ -776,6 +800,15 @@ module aptos_framework::fungible_asset {
             event::destroy_handle(withdraw_events);
             event::destroy_handle(frozen_events);
         };
+
+        if (features::is_fa_mint_burn_events_enabled()) {
```
Comment (igor-aptos) [inline aptos-move/framework/aptos-framework/sources/fungible_asset.move:804]: for all the places, let's also check it is not APT

maybe create a local functions here should_emit_mint_burn_events(asset_addr)

to check both feature flag and whether it is not APT

**Review by igor-aptos — CHANGES_REQUESTED** [review]

**Pull request closed without merging**